File: src/main/java/io/github/marcuscastelo/invtweaks/registry/InvTweaksBehaviorRegistry.java
package io.github.marcuscastelo.invtweaks.registry;

import io.github.marcuscastelo.invtweaks.InvTweaksOperationInfo;
import io.github.marcuscastelo.invtweaks.api.ScreenSpecification;
import io.github.marcuscastelo.invtweaks.client.behavior.InvTweaksVanillaGenericBehavior;
import io.github.marcuscastelo.invtweaks.client.behavior.InvTweaksVanillaMerchantBehavior;
import io.github.marcuscastelo.invtweaks.client.behavior.InvTweaksVanillaPlayerBehaviour;
import net.minecraft.screen.*;

import java.util.HashMap;

public class InvTweaksBehaviorRegistry {
    public static HashMap<Class<? extends ScreenHandler>, ScreenSpecification> screenBehaviorMap = new HashMap<>();

    public static ScreenSpecification register(ScreenSpecification screenSpecification) {
        if (screenBehaviorMap.containsKey(screenSpecification.getHandlerClass())) throw new IllegalArgumentException("Screen " + screenSpecification.getHandlerClass() + " is already registered");
        screenBehaviorMap.put(screenSpecification.getHandlerClass(), screenSpecification);
        return screenSpecification;
    }

    public static ScreenSpecification getScreenInfo(Class<? extends ScreenHandler> screenHandlerClass) {
        if (!isScreenSupported(screenHandlerClass)) throw new IllegalArgumentException("Screen " + screenHandlerClass + " is not supported");
        return screenBehaviorMap.get(screenHandlerClass);
    }

    public static void executeOperation(Class<? extends ScreenHandler> screenHandlerClass, InvTweaksOperationInfo operationInfo) throws IllegalArgumentException{
        if (!isScreenSupported(screenHandlerClass)) throw new IllegalArgumentException("Screen "  + screenHandlerClass + " doesn't have a behavior");
        operationInfo.type.asOperationExecutor(screenBehaviorMap.get(screenHandlerClass).getInvTweaksBehavior()).execute(operationInfo);
    }

    public static boolean isScreenSupported(Class<? extends ScreenHandler> screenHandlerClass) {
        return screenBehaviorMap.containsKey(screenHandlerClass);
    }

    public static ScreenSpecification.Builder createScreenInfoBuilder(Class<? extends ScreenHandler> handlerClass) {
        return new ScreenSpecification.Builder(handlerClass);
    }

    public static ScreenSpecification.Builder createVanillaGenericScreenInfoBuilder(Class<? extends ScreenHandler> handlerClass) {
        return createScreenInfoBuilder(handlerClass).setBehavior(new InvTweaksVanillaGenericBehavior());
    }

    public static ScreenSpecification buildDefaultedVanillaGenericScreenInfo(Class<? extends ScreenHandler> handlerClass) {
        return createVanillaGenericScreenInfoBuilder(handlerClass).build();
    }

    static {
        register(buildDefaultedVanillaGenericScreenInfo(GenericContainerScreenHandler.class));
        register(buildDefaultedVanillaGenericScreenInfo(ShulkerBoxScreenHandler.class));
        register(buildDefaultedVanillaGenericScreenInfo(CraftingScreenHandler.class));
        register(buildDefaultedVanillaGenericScreenInfo(BrewingStandScreenHandler.class));
        register(buildDefaultedVanillaGenericScreenInfo(BeaconScreenHandler.class));
        register(buildDefaultedVanillaGenericScreenInfo(HopperScreenHandler.class));
        register(buildDefaultedVanillaGenericScreenInfo(Generic3x3ContainerScreenHandler.class));
        register(buildDefaultedVanillaGenericScreenInfo(FurnaceScreenHandler.class));
        register(buildDefaultedVanillaGenericScreenInfo(SmokerScreenHandler.class));
        register(buildDefaultedVanillaGenericScreenInfo(BlastFurnaceScreenHandler.class));
        register(buildDefaultedVanillaGenericScreenInfo(EnchantmentScreenHandler.class));
        register(buildDefaultedVanillaGenericScreenInfo(AnvilScreenHandler.class));
        register(buildDefaultedVanillaGenericScreenInfo(CartographyTableScreenHandler.class));
        register(buildDefaultedVanillaGenericScreenInfo(LoomScreenHandler.class));
        register(buildDefaultedVanillaGenericScreenInfo(SmithingScreenHandler.class));
        register(buildDefaultedVanillaGenericScreenInfo(GrindstoneScreenHandler.class));
        register(buildDefaultedVanillaGenericScreenInfo(HorseScreenHandler.class));


        register(createScreenInfoBuilder(PlayerScreenHandler.class).setBehavior(new InvTweaksVanillaPlayerBehaviour()).setPlayerInvTotalSize(37).build());
        register(createScreenInfoBuilder(MerchantScreenHandler.class).setBehavior(new InvTweaksVanillaMerchantBehavior()).build());
    }

}

File: src/main/kotlin/io/github/marcuscastelo/invtweaks/registry/InvTweaksBehaviorRegistry.kt
package io.github.marcuscastelo.invtweaks.registry

import io.github.marcuscastelo.invtweaks.InvTweaksMod
import io.github.marcuscastelo.invtweaks.behavior.*
import io.github.marcuscastelo.invtweaks.operation.OperationInfo
import io.github.marcuscastelo.invtweaks.operation.OperationResult
import io.github.marcuscastelo.invtweaks.operation.OperationResult.Companion.pass
import io.github.marcuscastelo.invtweaks.util.ChatUtils.warnPlayer
import net.minecraft.client.gui.screen.ingame.CreativeInventoryScreen
import net.minecraft.client.gui.screen.ingame.CreativeInventoryScreen.CreativeScreenHandler
import net.minecraft.screen.*

object InvTweaksBehaviorRegistry {
    private var behaviorMap = mutableMapOf<Class<out ScreenHandler>, IInvTweaksBehavior>()
    val behaviors: Map<Class<out ScreenHandler>, IInvTweaksBehavior> get() = behaviorMap

    private val DEFAULT_BEHAVIOR = InvTweaksVanillaGenericBehavior()

    fun register(handlerClass: Class<out ScreenHandler>, behavior: IInvTweaksBehavior = InvTweaksVanillaGenericBehavior()): IInvTweaksBehavior {
        require(!behaviorMap.containsKey(handlerClass)) { "Screen $handlerClass is already registered" }
        behaviorMap[handlerClass] = behavior
        return behavior
    }

    fun executeOperation(screenHandlerClass: Class<out ScreenHandler>, operationInfo: OperationInfo): OperationResult {
        val behavior = behaviorMap[screenHandlerClass] ?: run {
            warnPlayer("Screen $screenHandlerClass doesn't have a behavior");
            DEFAULT_BEHAVIOR
        }

        val executor = operationInfo.type.asOperationExecutor(behavior)

        return executor?.execute(operationInfo) ?: run {
            InvTweaksMod.LOGGER.warn("<InvTweaksBehaviorRegistry> Operation " + operationInfo.type + " is not supported by " + behavior.javaClass)
            pass("Operation " + operationInfo.type + " is not supported by " + behavior.javaClass)
        }
    }

    fun isScreenRegistered(screenHandlerClass: Class<out ScreenHandler>): Boolean {
        return behaviorMap.containsKey(screenHandlerClass)
    }

    init {
        register(GenericContainerScreenHandler::class.java)
        register(ShulkerBoxScreenHandler::class.java)
        register(BrewingStandScreenHandler::class.java)
        register(BeaconScreenHandler::class.java)
        register(HopperScreenHandler::class.java)
        register(Generic3x3ContainerScreenHandler::class.java)
        register(FurnaceScreenHandler::class.java)
        register(SmokerScreenHandler::class.java)
        register(BlastFurnaceScreenHandler::class.java)
        register(AnvilScreenHandler::class.java)
        register(CartographyTableScreenHandler::class.java)
        register(LoomScreenHandler::class.java)
        register(SmithingScreenHandler::class.java)
        register(GrindstoneScreenHandler::class.java)
        register(HorseScreenHandler::class.java)
        register(CraftingScreenHandler::class.java)
        register(StonecutterScreenHandler::class.java)
        register(CreativeScreenHandler::class.java)

        //Survival Inventory
        register(PlayerScreenHandler::class.java, InvTweaksVanillaPlayerBehaviour())

        //Merchant behaviour
        register(MerchantScreenHandler::class.java, InvTweaksVanillaMerchantBehavior())

        //Enchantment behaviour
        register(EnchantmentScreenHandler::class.java, EnchantingTableBehavior())

    }
}